azure/credential_cache.go

```diff
@@ -0,0 +1,159 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azure
+
+import (
+	"sync"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+)
+
+type credentialCache struct {
+	mut         *sync.Mutex
+	cache       map[credentialCacheKey]azcore.TokenCredential
+	credFactory credentialFactory
+}
+
+type credentialFactory interface {
+	newClientSecretCredential(tenantID string, clientID string, clientSecret string, opts *azidentity.ClientSecretCredentialOptions) (azcore.TokenCredential, error)
+	newClientCertificateCredential(tenantID string, clientID string, clientCertificate []byte, clientCertificatePassword []byte, opts *azidentity.ClientCertificateCredentialOptions) (azcore.TokenCredential, error)
+	newManagedIdentityCredential(opts *azidentity.ManagedIdentityCredentialOptions) (azcore.TokenCredential, error)
+	newWorkloadIdentityCredential(opts *azidentity.WorkloadIdentityCredentialOptions) (azcore.TokenCredential, error)
+}
+
+// CredentialType represents the auth mechanism in use.
+type CredentialType int
+
+const (
+	// CredentialTypeClientSecret is for Service Principals with Client Secrets.
+	CredentialTypeClientSecret CredentialType = iota
+	// CredentialTypeClientCert is for Service Principals with Client certificates.
+	CredentialTypeClientCert
+	// CredentialTypeManagedIdentity is for Managed Identities.
+	CredentialTypeManagedIdentity
+	// CredentialTypeWorkloadIdentity is for Workload Identity.
+	CredentialTypeWorkloadIdentity
+)
+
+type credentialCacheKey struct {
+	authorityHost  string
+	credentialType CredentialType
+	tenantID       string
+	clientID       string
+}
+
+// NewCredentialCache creates a new, empty CredentialCache.
+func NewCredentialCache() CredentialCache {
+	return &credentialCache{
+		mut:         new(sync.Mutex),
+		cache:       make(map[credentialCacheKey]azcore.TokenCredential),
+		credFactory: azureCredentialFactory{},
+	}
+}
+
+func (c *credentialCache) GetOrStoreClientSecret(tenantID, clientID, clientSecret string, opts *azidentity.ClientSecretCredentialOptions) (azcore.TokenCredential, error) {
+	return c.getOrStore(
+		credentialCacheKey{
+			authorityHost:  opts.Cloud.ActiveDirectoryAuthorityHost,
+			credentialType: CredentialTypeClientSecret,
+			tenantID:       tenantID,
+			clientID:       clientID,
+		},
+		func() (azcore.TokenCredential, error) {
+			return c.credFactory.newClientSecretCredential(tenantID, clientID, clientSecret, opts)
+		},
+	)
+}
+
+func (c *credentialCache) GetOrStoreClientCert(tenantID, clientID string, cert, certPassword []byte, opts *azidentity.ClientCertificateCredentialOptions) (azcore.TokenCredential, error) {
+	return c.getOrStore(
+		credentialCacheKey{
+			authorityHost:  opts.Cloud.ActiveDirectoryAuthorityHost,
+			credentialType: CredentialTypeClientCert,
+			tenantID:       tenantID,
+			clientID:       clientID,
```

Comment:
@bryan-cox Do you happen to know if adding or changing a certificate for a Service Principal keeps the existing client ID? i.e. Should the certificate data itself be a part of this key used to cache it?

Reply:
Yeah it keeps the existing client ID. When I've been testing with SP w/cert, I have just been rotating the cert in Azure key vault so even the cert name stays the same, only the certificate contents change. Since the certificate data changes, it makes sense to include it as part of the key used to cache.

Reply:
Thanks! I'll plan to add that here.

/hold

Reply:
Added secrets to the key.

/hold cancel